ls1intum/Artemis
# [Bugfix] Maintain CommitState Consistency

### Checklist

### Motivation and Context
See issue #1854 
### Description
This was already started in #1924 but after major code editor refactoring I decided it was best to start fresh.
- When the editor state changes to `EditorState.UNSAVED_CHANGES`, the commit state also changes to `CommitState.UNCOMMITTED_CHANGES`.
- Because the commit state is now dependent on the unsaved changes, we also need to consider the possibility of the user changing a file while a commit is undertaken. In that case we must set the commit state to `CommitState.UNCOMMITTED_CHANGES`. This is done inside `ngOnChanges` because the editor state is updated in another component (`CodeEditorContainerComponent#unsavedFiles`) which maintains the unsaved file status. The alternative to utilizing `ngOnChanges` would be to use busy waiting which would reduce the performance. When we do a commit, if there are unsaved changes we first do a save and then the commit. If we do not wait for the 'save' to settle we run into race conditions which lead to inconsistent results. 

### Steps for Testing
1. Log in to Artemis
2. Start a programming exercise with the code editor enabled
3. Make some changes. 
    The commit state should become `CommitState.UNCOMMITTED_CHANGES`.
4. Press save. 
5. Press commit. 
    The commit state should become `CommitState.CLEAN`
6. Make more changes and press commit immediately. 
    The commit state should become `CommitState.CLEAN`
7. Make more changes and press commit. While it is committing, make more changes and observe the status. 
    It should become: `CommitState.UNCOMMITED_CHANGES`. Unless the auto-save kicked in, the editor state should also be  `EditorState.UNSAVED_CHANGES`.
8. Make more changes. and press commit. As soon as it is possible press refresh files. 
    The editor and commit state should be  `EditorState.CLEAN` and  `CommitState.CLEAN` respectively.

### Test Coverage
CodeEditorContainerComponent => 86%
CodeEditorActions => 78.5%
### Screenshots
![image](https://user-images.githubusercontent.com/16407766/87837966-b1fc3680-c895-11ea-9c65-7d05612e756b.gif)
